Add support for PHP 8.1

arueckauer opened the pull request:
|    Q          |   A
|-------------- | ------
| QA            | yes

### Description

PHP 8.1 support.

This also fixes known incompatibilites with PHP 8.1 described in #35 and #36.

arueckauer committed:
Add default value to property projectRoot of ComponentInstaller preventing deprecation notice of passing null to is_dir

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove unnecessary value assignment of projectRoot which causes an error by static code analyzer

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Initialize mocks in test method without the need to rely or alter setup()

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add argument type declaration to ensure type safety and null is never passed

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add PHP 8.1 support

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Fix incompatibility with ArrayAccess, Countable and IteratorAggregate

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

boesing commented:
To make `composer install` pass, you need to add `.laminas-ci.json` to the project root.

```json
{
    "ignore_php_platform_requirements": {
        "8.1": true
    }
}
```

boesing commented on src/Collection.php:
```diff
@@ -19,6 +19,8 @@
 use function iterator_to_array;
 use function sprintf;
 
+use const ReturnTypeWillChange;
```
Thats what I meant in slack, you have to manually `use` the Attribute here (which is *not* a constant).

Codestyle will comply about the unused `ReturnTypeWillChange` `use`-Statement and will remove it when applying `cs-fix`.

To avoid that, you can add a class docblock:

```php
/**
 * @use ReturnTypeWillChange
 */
class Collection […]
```

arueckauer replied:
Fixed.

arueckauer committed:
Exclude 8.1 in CI

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

boesing commented on composer.lock:
```diff
@@ -4672,37 +4750,38 @@
         {
             "name": "vimeo/psalm",
-            "version": "4.6.4",
+            "version": "4.10.0",
```
You've updated *everything* and thus also updated `vimeo/psalm` which now leads to psalm errors.
Thats why you should only use `composer update --lock` if you change something in the `composer.json` which is not actually installing anything (like config stuff, non-installable dependencies like extensions, php versions, e.g.).

If you have to bump packages to be compatible with PHP 8.1, you might want to require these new package versions explicitly.

You can also bump psalm but then fixing the errors should be mandatory.  If you don't want to fix them, do not update psalm.

arueckauer replied:
Totally overlooked that. Reverted accidental dependencies upgrade.

arueckauer committed:
Revert upgrade of dependencies

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Upgrade laminas/laminas-coding-standard to 2.2.1

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Fix import of ReturnTypeWillChange

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Upgrade mikey179/vfsstream to 1.6.10 for PHP8.1 compatibility

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Upgrade lowest version of composer/composer to 1.6.10 for PHP 8.1 compatibility

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove lowest version of composer/composer due to incompatibility with PHP 8.1
Upgrade composer/composer to 2.0.14 to add compatibility with new key format of GitHub tokens

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Upgrade composer/composer to 2.1.9 and vimeo/psalm to 4.10 for PHP 8.1 compatibility

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer commented:
As version `^2.0` of `composer/xdebug-handler` is required, `vimeo/psalm` needs to be updated to `^4.8`. Since from there to `^4.10` only one more error is thrown, I did the update to the latest version.

Would be happy to get some hints on the four argument errors. With the unreferenced param and `@internal` errors, I am not sure, how to fix it or if it could be added to the baseline.

boesing commented:
Since the psalm errors are not related to the diff, there is no way of adding comments to those lines by using github.

I will list the issues below and add suggestions on how to fix these:


> Argument 1 of Laminas\ComponentInstaller\ComponentInstaller::mapAutoloaders expects array<string, array<int|string, list<string>|string>>, parent type array{classmap?: list<string>, files?: list<string>, psr-0?: array<string, array<array-key, string>|string>, psr-4?: array<string, array<array-key, string>|string>} provided (see https://psalm.dev/194)

Okay, to make psalm happy, we could ~import the psalm/phpstan type (which is used in the `PackageInterface` of composer)~ copy and paste the return type from composer (sadly they did not used a [local type](https://phpstan.org/writing-php-code/phpdoc-types#local-type-aliases) for that - might be worth changing that, not sure if thats something composer wants to implement) and then change the type for `mapAutoloaders`. 
https://github.com/composer/composer/blob/cb1e2482580beaf3e511278dd629ded599cddc1f/src/Composer/Package/PackageInterface.php#L300

---

> Param $packages is never referenced in this method (see https://psalm.dev/188)

> Param $type is never referenced in this method (see https://psalm.dev/188)

> Param $key is never referenced in this method (see https://psalm.dev/188)

> Param $injector is never referenced in this method (see https://psalm.dev/188)

Just change unused params to `$_` - this should be fine for both the coding standard and psalm. any variable prefixed with `_` is being ignored if it comes to unused parameters, so if there are more than one unused argument within a function, you can also add `$_1`, e.g.

---

> Unable to determine the type that $allowedTypes is being assigned to (see https://psalm.dev/032)

This could be fixed by adding generics to the `Collection` implementation.
https://psalm.dev/docs/annotating_code/templated_annotations/#param-class-stringt

---

> Argument 1 of Composer\Repository\InstalledRepository::__construct expects array<array-key, Composer\Repository\RepositoryInterface>, array{Composer\Repository\RootPackageRepository|null, Composer\Repository\InstalledRepositoryInterface, Composer\Repository\PlatformRepository} provided (see https://psalm.dev/004)


So it seems that we somehow have an annotation which is `null` for the `packageRepository` property. Either remove that `null` and ensure that its definitely not set to `null` at any point (e.g. by using `webmozart/assert`) or use `webmozart/assert` before passing that argument to the `InstalledRepository`

---

> Constructor PHPUnit\Framework\ExpectationFailedException::__construct is internal to PHPUnit but called from LaminasTest\ComponentInstaller (see https://psalm.dev/175)

> Constructor PHPUnit\Framework\ExpectationFailedException::__construct is internal to PHPUnit but called from LaminasTest\ComponentInstaller (see https://psalm.dev/175)

> Constructor PHPUnit\Framework\ExpectationFailedException::__construct is internal to PHPUnit but called from LaminasTest\ComponentInstaller (see https://psalm.dev/175)


Just change this to `self::fail()` rather than throwing an exception here.

boesing commented:
I definitely want to deep-dive more into this as I did plenty of work to make both composer v1 and v2 compatible with this component. dropping v1 because of a minor PHP version should be avoided at any cost.

Could you probably re-add composer v1 as a dependency (and ensure that it will be installed when using `composer update --prefer-lowest`)? I'd love to see which tests are failing. If only PHP 8.1 tests are failing when using `lowest` deps, I probably would be fine with adding that test to `exclude` within `.laminas-ci.json`.

But maybe this component needs also a different CI pipeline so we have better coverage for the dedicated composer versions. I'll think about that somewhen in the next weeks. 

Please ping me here or on slack if you think I might've forgotten this (which could probably be true).

arueckauer committed:
Replace usage of internal ExpectationFailedException with self::fail()

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add MissingConstructor suppression for properties which will be initialized in setUp()

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Replace usage of internal ExpectationFailedException with self::fail()

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

boesing commented on test/ConfigDiscoveryTest.php:
```diff
@@ -19,6 +19,9 @@
 use function sprintf;
 
+/**
+ * @psalm-suppress MissingConstructor
```
if you add the psalm phpunit plugin, this wont be needed.

snapshotpl commented on composer.json:
```diff
@@ -25,18 +25,18 @@
+        "laminas/laminas-coding-standard": "2.2.1",
         "malukenho/docheader": "^0.1.6",
-        "mikey179/vfsstream": "^1.6.7",
+        "mikey179/vfsstream": "1.6.10",
```
Why without caret?

arueckauer replied:
Caret restored.

snapshotpl commented on src/ConfigDiscovery/AbstractDiscovery.php:
```diff
@@ -33,10 +33,8 @@ abstract class AbstractDiscovery implements DiscoveryInterface
-     * @param string $projectDirectory
      */
-    public function __construct($projectDirectory = '')
+    public function __construct(string $projectDirectory = '')
```
This change introduce compatibility break, because changes public interface

arueckauer replied:
I see your point.

IIRC I discussed this with @boesing in chat. The `@param` type hint states parameter is of type string. If implementation is breaking, it is due to a misuse.

Is that correct, @boesing?

boesing replied:
Sadly, `__construct` has some [special behavior](https://3v4l.org/73N5Q) if it comes to arguments. So this is actually no BC break until the type-hint matches what we do state in the `@var` annotation.

(sidenote: imho `laminas-component-installer` is not a library and thus, semver does not really make sense here at all. But actually, we do not differ between libraries and components like this (composer plugin))

boesing replied:
fun-fact: this constructor argument is only added for unit tests so we can pass virtual filesystem directory

arueckauer committed:
Add psalm/plugin-phpunit which makes MissinConstructor suppress directives for Psalm unnecessary

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Restore caret for version constraint of mikey179/vfsstream

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

weierophinney committed:
feat: bump minimum supported PHP version

- Bumps minimum supported PHP version to 7.4
- Updates to laminas-coding-standard 2.3 series
- Bumps minimum supported PHPUnit version to 9.5.5
- Updates how package replaces zend-component-installer by converting "replace" rule to a "conflict" rule against all versions, and removing the laminas-zendframework-bridge.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: apply laminas-coding-standard 2.3 rules and ensure PHPUnit converts deprecations to exceptions

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: allow proper working under PHP 8.1

- Import the `ReturnTypeWillChange` attribute (`@use` annotation does not work for engine purposes)
- All dependencies are now installable under PHP 8.1, so no need to ignore platform requirements :tada:

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: Psalm issues

The majority of these are due to third-party packages not providing specific enough of return type hints, leading to "mixed assignment" and similar.
Others are due to usage of things like filter/reduce/walk whereby we are operating on a subset of all arguments, but the last one is required for the operation (unused closure argument).

I did refactor the PackageProviderDetectionFactory to be more clear about what it provides, and to ensure it never passes a null argument.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
